ocaml/stdlib/mutex.ml

```diff
@@ -23,9 +23,10 @@ external unlock: t -> unit = "caml_ml_mutex_unlock"
 external reraise : exn -> 'a = "%reraise"
 
 (* cannot inline, otherwise flambda might move code around. *)
+(* CR mshinwell: [@inline never] should not be needed on flambda-backend *)
 let[@inline never] protect m f =
   lock m;
-  match f() with
+  match Sys.with_async_exns f with
```

Comment:
Reverting

Reply:
> It also means that async exceptions occuring while a mutex is locked leave the mutex unlocked but the data it protects in an unknown state.

Isn't that also true with regular exceptions ? Why would it be more problematic with asynchronous ones ?

Reply:
Right, you have to be careful with regular exceptions here too. But at least it is possible to be careful in this case, since the places that can raise regular exceptions are visible in the source.

Reply:
Yes, but in the world before this PR `Mutex.protect` guarantees that it will leave the mutex unlocked, whatever happened (including asynchronous exceptions). If we don't add `Sys.with_async_exns` then we change this property.